### PR TITLE
Give the user more flexibility

### PR DESCRIPTION
**Pull request by Konubinix**
Those commits don't change the default behavior of uzbl. They attempt to replace hard coded behavior by configurable ones.

### CODE REVIEW

File: examples/config/config [hunk 1]
```diff
@@ -6,6 +6,7 @@
 # common directory locations
 set prefix      @(echo $UZBL_PREFIX)@
 set data_home   @(echo $XDG_DATA_HOME/uzbl)@
+set data_dirs   @(echo $(echo ${XDG_DATA_DIRS}|sed 's%:%/uzbl:%g')/uzbl)@
```
Comment (mathstuf) [inline examples/config/config]: This misses the last directory in the list.

File: bin/uzbl-browser.in [hunk 2]
```diff
@@ -15,6 +15,9 @@ EXAMPLES="$UZBL_PREFIX/share/uzbl/examples"
 XDG_DATA_HOME="${XDG_DATA_HOME:-$HOME/.local/share}"
 export XDG_DATA_HOME
 
+XDG_DATA_DIRS="${XDG_DATA_DIRS:-/usr/local/share/:/usr/share/}"
```
Comment (mathstuf) [inline bin/uzbl-browser.in]: Remove trailing slashes please.

File: uzbl/plugins/cookies.py [hunk 3]
```diff
@@ -163,8 +163,10 @@ def delete_cookie(self, rkey, key):
         os.umask(curmask)
 
 xdg_data_home = os.environ.get('XDG_DATA_HOME', os.path.join(os.environ['HOME'], '.local/share'))
-DefaultStore = TextStore(os.path.join(xdg_data_home, 'uzbl/cookies.txt'))
-SessionStore = TextStore(os.path.join(xdg_data_home, 'uzbl/session-cookies.txt'))
+UZBL_COOKIE_FILE = os.environ.get("UZBL_COOKIE_FILE", os.path.join(xdg_data_home, 'uzbl/cookies.txt'))
+UZBL_SESSION_COOKIE_FILE = os.environ.get("UZBL_SESSION_COOKIE_FILE", os.path.join(xdg_data_home, 'uzbl/session-cookies.txt'))
+DefaultStore = TextStore(UZBL_COOKIE_FILE)
+SessionStore = TextStore(UZBL_SESSION_COOKIE_FILE)
```
Comment (mathstuf) [inline uzbl/plugins/cookies.py]: Looks good, but remove the execute permissions you added to the file.

File: examples/config/config (hunk 1 above)
Comment (Konubinix) [inline examples/config/config]: Actually, adding /uzbl to the last directory is the reason why I had to put two nested echo commands.

If XDG_DATA_DIRS=a:b:c, then
echo $(echo ${XDG_DATA_DIRS}|sed 's%:%/uzbl:%g')/uzbl
Will correctly output a/uzbl:b/uzbl:c/uzbl

File: uzbl/plugins/cookies.py (hunk 3 above)
Comment (Konubinix) [inline uzbl/plugins/cookies.py]: Done.

Thanks

File: examples/config/config (hunk 1 above)
Comment (mathstuf) [inline examples/config/config]: Ah, indeed.

File: examples/config/config [hunk 4]
```diff
@@ -516,7 +517,7 @@ set ssl_policy fail
 
 # === Post-load misc commands ================================================
 spawn_sync_exec @scripts_dir/load_cookies.sh
-spawn_sync_exec @scripts_dir/load_cookies.sh @data_home/session-cookies.txt
+spawn_sync_exec @scripts_dir/load_cookies.sh @(echo "${UZBL_SESSION_COOKIE_FILE}")@
```
Comment (mathstuf) [inline examples/config/config]: Hrm. This doesn't work if this isn't set in the environment :( . Maybe a variable should be used that cookies.py sets? Alternatively, pass a list of paths to `load_cookies.sh` and load the first (all?) of them.

Comment (Konubinix) [inline examples/config/config]: It suppose the script examples/data/scripts/util/uzbl-dir.sh has been sourced somewhere.

The same assumption is made at the beginning of this file. The setting relying on
prefix, data_home, data_dirs, cache_home and config_home don't work either if
the XDG environment variables are not set.

For instance, the ctrl-m binding (adding a bookmark) won't work if data_home is not correctly set, which is the case if XDG_DATA_HOME is not correctly set, which could be the case if examples/data/scripts/util/uzbl-dir.sh is not sourced.

Comment (mathstuf) [inline examples/config/config]: On Wed, Feb 11, 2015 at 08:43:39 -0800, Samuel Loury wrote:
> It suppose the script examples/data/scripts/util/uzbl-dir.sh has been
> sourced somewhere.

The uzbl-dir.sh code is meant for use inside of the scripts; this is the
configuration file which can't assume its use (the user would need to
run it for it to take effect). The configuration file is also supposed
to work with uzbl-browser, uzbl-tabbed, or any other tool when uses
uzbl-core directly.

> The same assumption is made at the beginning of this file. The setting relying on
> prefix, data_home, data_dirs, cache_home and config_home don't work either if
> the XDG environment variables are not set.
> 
> For instance, the ctrl-m binding (adding a bookmark) won't work if
> data_home is not correctly set, which is the case if XDG_DATA_HOME is
> not correctly set, which could be the case if
> examples/data/scripts/util/uzbl-dir.sh is not sourced.

The `_HOME` variables are handled in multiple places. First,
uzbl-browser sets them up for itself and uzbl-core. uzbl-core will also
ensure that the `_HOME` variables exist and will use the `_DIRS`
variants for finding files as a fallback, but it won't set those.

Comment (Konubinix) [inline examples/config/config]: I understand. Thank you for taking the time to explain :-).

I will try to think of and submit a solution to this issue.

File: examples/config/config [hunk 5]
```diff
@@ -516,7 +517,7 @@ set ssl_policy fail
 
 # === Post-load misc commands ================================================
 spawn_sync_exec @scripts_dir/load_cookies.sh
-spawn_sync_exec @scripts_dir/load_cookies.sh @data_home/session-cookies.txt
+spawn_sync_exec @scripts_dir/load_cookies.sh @(echo "${UZBL_SESSION_COOKIE_FILE:-@data_home/session-cookies.txt}")@
```
Comment (Konubinix) [inline examples/config/config]: Here, the UZBL_SESSION_COOKIE_FILE environment variable is used.

It it is not set, the set command falls back to the same default value than used in the cookies plugins and the scripts.

**Comment by Konubinix** [comment]
Do you think those changes could be applied?

**Comment by mathstuf** [comment]
Yeah, it looks good to me. I'll pull it when I get around to doing more uzbl stuff (hopefully Sunday, but I can make no guarantees).

**Pull request merged**